### PR TITLE
Dictionary and Array encoding

### PR DESCRIPTION
**Pull request by adam-fowler**
For XML and query encoding there are multiple ways to encode dictionaries and arrays. This PR is an attempt to pull all of these together. The idea is to add an additional member to AWSShapeMember which defines how a collection is encoded. The collection encoding is defined by an enum with associated values. 

Included in this PR are the changes to the CodeGenerator and the re-generated shape files.

More details about this can be found in it's sibling PR in the aws-sdk-swift-core repository https://github.com/swift-aws/aws-sdk-swift-core/pull/86

This PR fixes https://github.com/swift-aws/aws-sdk-swift/issues/74

### CODE REVIEW

File: Sources/CodeGenerator/AWSService.swift [hunk 1]
```diff
@@ -155,20 +156,41 @@ struct AWSService {
...
-                    location = Location(key:flattenedLocationName, json: shapeJSON["member"])
+                var encoding : CollectionEncoding? = nil
+                // xml and query encoding needs collection encoding information
+                switch serviceProtocol.type {
+                case .query, .restxml:
+                    encoding = CollectionEncoding(json: shapeJSON)
+                default:
+                    break
+                }
+                // if list is flattened
```
[the remark this answers is not in the document]
Comment (Yasumoto) [reply Sources/CodeGenerator/AWSService.swift]: With these three, I tried to expand them out for folks in the future to have more context. Please feel free to reject or expand if I misunderstood!

File: Sources/CodeGenerator/AWSService.swift [hunk 2]
```diff
@@ -155,20 +156,41 @@ struct AWSService {
...
+                if let flattenedLocationName = shapeJSON["member"]["locationName"].string, shapeJSON["flattened"].bool == true {
+                    location = Location(json: shapeJSON["member"])
                     locationName = flattenedLocationName
                 }
+                // if this instance of the collection change the encoding to the equivalent flattened version
```
[the remark this answers is not in the document]
Comment (adam-fowler) [reply Sources/CodeGenerator/AWSService.swift]: done

File: Sources/CodeGenerator/AWSService.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (adam-fowler) [reply Sources/CodeGenerator/AWSService.swift]: done

File: Sources/CodeGenerator/AWSService.swift [hunk 3]
```diff
@@ -121,7 +121,7 @@ struct AWSService {
 
         case "list":
             let shapeJSON = apiJSON["shapes"][json["member"]["shape"].stringValue]
-            if let locationName = json["member"]["locationName"].string, json["flattened"].bool != true {
+            /*if let locationName = json["member"]["locationName"].string, json["flattened"].bool != true {
```
[the remark this answers is not in the document]
Comment (adam-fowler) [reply Sources/CodeGenerator/AWSService.swift]: done

**Comment by adam-fowler** [comment]
@Yasumoto can you post the XML you are receiving, so I can have a look at why it isn't decoding

**Comment by adam-fowler** [comment]
@Yasumoto Code generator was only adding encoding info for xml and query services. Ec2 has its own special service type. Sorted that but there seems to be a query construction issue now. Hopefully will get a chance to look at it before the end of the day.

**Comment by Yasumoto** [comment]
Great sleuthing 🕵️‍♂️!! Thank you @adam-fowler

**Comment by adam-fowler** [comment]
@Yasumoto Ok looks like EC2 is a law unto itself. The model files do not agree with what the service requires. All lists in queries are flattened but this is not indicated in the model file. I fixed this by adding an additional flattenLists param to AWSShapeEncoder.query(). This defaults to false but the EC2 call sets this to true. Latest version of core code along with updated EC2_Shapes.swift file should work for you.

I could do this in the CodeGenerator but deciding on whether a structure is in a request or a result would be a pain. Might return to this later though.

**Pull request merged**